lib/Conversion/HWToBTOR2/HWToBTOR2.cpp

```diff
@@ -0,0 +1,817 @@
+  static constexpr StringLiteral orStr = "or";
+  static constexpr StringLiteral xorStr = "xor";
+  static constexpr StringLiteral sllStr = "sll";
+  static constexpr StringLiteral srlStr = "srl"; // a.k.a. unsigned right shift
+  static constexpr StringLiteral sraStr = "sra"; // a.k.a. signed right shift
+  static constexpr StringLiteral sdivStr = "sdiv";
+  static constexpr StringLiteral udivStr = "udiv";
+  static constexpr StringLiteral smodStr = "smod";
+  static constexpr StringLiteral concatStr = "concat";
+  static constexpr StringLiteral notStr = "not";
+  static constexpr StringLiteral neqStr = "neq";
+  static constexpr StringLiteral hwNeqStr = "ne";
+  static constexpr StringLiteral iteStr = "ite";
+  static constexpr StringLiteral impliesStr = "implies"; // logical implication
+  static constexpr StringLiteral stateStr = "state";     // Register state
+  static constexpr StringLiteral nextStr = "next"; // Register state transition
+  static constexpr StringLiteral badStr = "bad";
+  static constexpr StringLiteral constraintStr = "constraint";
+  static constexpr StringLiteral wsStr = " ";  // WhiteSpace
+  static constexpr StringLiteral nlStr = "\n"; // NewLine
+
+  // Constants used during the conversion
+  static constexpr size_t noLID = -1UL;
+  static constexpr int64_t noWidth = -1L;
+
+  /// Field helper functions
+
+  // Checks if an operation was declared
+  // If so, its lid will be returned
+  // Otherwise -1 will be returned
+  size_t getOrCreateOpLID(Operation *op) {
+    if (op == nullptr)
+      return noLID;
+
+    // Look for the original operation declaration
+    // Make sure that wires are considered when looking for an lid
+    Operation *defOp = getOpAlias(op);
+    if (opLIDMap.contains(defOp)) {
+      return opLIDMap[defOp];
+    }
+
+    // Create a new entry with the current LID
+    // if the op was not yet registered
+    opLIDMap[defOp] = lid;
+
+    return opLIDMap.at(defOp);
+  }
+
+  // Forces a new LID to be associated to the given op
+  size_t updateOpLID(Operation *op) {
+    if (op == nullptr)
+      return noLID;
+
+    opLIDMap[op] = lid;
+    return lid;
+  }
+
+  // Checks if an operation was declared
+  // If so, its lid will be returned
+  // Otherwise -1 will be returned
+  size_t getOrCreateOpLID(Value op) {
+    // Check for an operation alias
+    Operation *defOp = getOpAlias(op.getDefiningOp());
+
+    if (auto it = opLIDMap.find(defOp); it != opLIDMap.end())
+      return it->second;
+
+    // Check for special case where op is actually a port
+    // To do so, we start by checking if our operation is a block argument
+    if (BlockArgument barg = dyn_cast<BlockArgument>(op)) {
+      // Extract the block argument index and use that to get the line number
+      size_t argIdx = barg.getArgNumber();
+
+      // Check that the extracted argument is in range before using it
+      if (auto it = inputLIDs.find(argIdx); it != inputLIDs.end())
```

Comment:
this can be folded into later code.  A single lookup should be sufficient for all code paths.